// sanity-plugin-media/dist/styled/react-select/creatable.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactSelectComponents = exports.reactSelectStyles = void 0;
var color_1 = require("@sanity/color");
var icons_1 = require("@sanity/icons");
var ui_1 = require("@sanity/ui");
var react_1 = __importDefault(require("react"));
var react_select_1 = require("react-select");
var react_virtuoso_1 = require("react-virtuoso");
var themeDarkPrimaryBlue = (_d = (_c = (_b = (_a = ui_1.studioTheme === null || ui_1.studioTheme === void 0 ? void 0 : ui_1.studioTheme.color) === null || _a === void 0 ? void 0 : _a.dark) === null || _b === void 0 ? void 0 : _b.primary) === null || _c === void 0 ? void 0 : _c.spot) === null || _d === void 0 ? void 0 : _d.blue;
var themeDarkPrimaryGray = (_h = (_g = (_f = (_e = ui_1.studioTheme === null || ui_1.studioTheme === void 0 ? void 0 : ui_1.studioTheme.color) === null || _e === void 0 ? void 0 : _e.dark) === null || _f === void 0 ? void 0 : _f.primary) === null || _g === void 0 ? void 0 : _g.spot) === null || _h === void 0 ? void 0 : _h.gray;
var themeRadius = ui_1.studioTheme === null || ui_1.studioTheme === void 0 ? void 0 : ui_1.studioTheme.radius;
var themeSpace = ui_1.studioTheme === null || ui_1.studioTheme === void 0 ? void 0 : ui_1.studioTheme.space;
exports.reactSelectStyles = {
    control: function (styles, _a) {
        var isDisabled = _a.isDisabled, isFocused = _a.isFocused;
        var boxShadow = "inset 0 0 0 1px " + color_1.hues.gray[900].hex;
        if (isFocused) {
            boxShadow = "inset 0 0 0 1px " + color_1.hues.gray[900].hex + ", 0 0 0 1px var(--card-bg-color), 0 0 0 3px var(--card-focus-ring-color) !important";
        }
        return __assign(__assign({}, styles), { background: isDisabled
                ? ui_1.studioTheme.color.dark.default.input.default.disabled.bg
                : 'transparent', color: 'white', border: 'none', borderRadius: themeRadius[1], boxShadow: boxShadow, minHeight: '35px', outline: 'none', transition: 'none', '&:hover': {
                boxShadow: "inset 0 0 0 1px " + ui_1.studioTheme.color.dark.default.input.default.hovered.border
            } });
    },
    indicatorsContainer: function (styles, _a) {
        var isDisabled = _a.isDisabled;
        return (__assign(__assign({}, styles), { opacity: isDisabled ? 0.25 : 1 }));
    },
    input: function (styles) { return (__assign(__assign({}, styles), { color: 'white', marginLeft: themeSpace[2] })); },
    menuList: function (styles) { return (__assign({}, styles)); },
    multiValue: function (styles, _a) {
        var isDisabled = _a.isDisabled;
        return (__assign(__assign({}, styles), { backgroundColor: themeDarkPrimaryGray, borderRadius: themeRadius[2], opacity: isDisabled ? 0.5 : 1 }));
    },
    multiValueRemove: function (styles) { return (__assign(__assign({}, styles), { paddingLeft: 0, '&:hover': {
            background: 'transparent',
            color: 'inherit'
        } })); },
    option: function (styles, _a) {
        var isFocused = _a.isFocused;
        return (__assign(__assign({}, styles), { backgroundColor: isFocused ? themeDarkPrimaryBlue : 'transparent', borderRadius: themeRadius[2], color: isFocused ? color_1.black.hex : 'inherit', padding: '4px 8px', '&:hover': {
                backgroundColor: themeDarkPrimaryBlue,
                color: color_1.black.hex
            } }));
    },
    placeholder: function (styles) { return (__assign(__assign({}, styles), { marginLeft: themeSpace[2] })); },
    valueContainer: function (styles) { return (__assign(__assign({}, styles), { marginBottom: themeSpace[0], marginLeft: themeSpace[1], marginTop: themeSpace[0], padding: 0 })); }
};
var DropdownIndicator = function (props) {
    return (react_1.default.createElement(react_select_1.components.DropdownIndicator, __assign({}, props),
        react_1.default.createElement(ui_1.Box, { paddingX: 2 },
            react_1.default.createElement(ui_1.Text, { size: 1 },
                react_1.default.createElement(icons_1.ChevronDownIcon, null)))));
};
var Menu = function (props) {
    return (react_1.default.createElement(react_select_1.components.Menu, __assign({}, props),
        react_1.default.createElement(ui_1.Card, { radius: 1, shadow: 2 }, props.children)));
};
var MenuList = function (props) {
    var children = props.children;
    var MAX_ROWS = 5;
    var OPTION_HEIGHT = 37;
    if (Array.isArray(children)) {
        var height = children.length > MAX_ROWS ? OPTION_HEIGHT * MAX_ROWS : children.length * OPTION_HEIGHT;
        return (react_1.default.createElement(react_virtuoso_1.Virtuoso, { className: "media__custom-scrollbar", itemContent: function (index) {
                var item = children[index];
                return react_1.default.createElement(Option, __assign({}, item.props));
            }, style: { height: height }, totalCount: children.length }));
    }
    return react_1.default.createElement(react_select_1.components.MenuList, __assign({}, props), children);
};
var MultiValueLabel = function (props) {
    return (react_1.default.createElement(ui_1.Box, { paddingLeft: 1, paddingRight: 0, paddingY: 1 },
        react_1.default.createElement(ui_1.Text, { size: 2, weight: "medium" },
            react_1.default.createElement(react_select_1.components.MultiValueLabel, __assign({}, props)))));
};
var MultiValueRemove = function (props) {
    return (react_1.default.createElement(react_select_1.components.MultiValueRemove, __assign({}, props),
        react_1.default.createElement(icons_1.CloseIcon, { color: "#1f2123" })));
};
var Option = function (props) {
    return (react_1.default.createElement(ui_1.Box, { paddingX: 1, paddingY: 1 },
        react_1.default.createElement(react_select_1.components.Option, __assign({}, props),
            react_1.default.createElement(ui_1.Flex, { align: "center" },
                props.data.__isNew__ && react_1.default.createElement(icons_1.AddIcon, { style: { marginRight: '3px' } }),
                props.children))));
};
exports.reactSelectComponents = {
    DropdownIndicator: DropdownIndicator,
    IndicatorSeparator: null,
    Menu: Menu,
    MenuList: MenuList,
    MultiValueLabel: MultiValueLabel,
    MultiValueRemove: MultiValueRemove,
    Option: Option
};
