// sanity-plugin-media/dist/components/TagsPanel/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var color_1 = require("@sanity/color");
var ui_1 = require("@sanity/ui");
var react_1 = __importDefault(require("react"));
var constants_1 = require("../../constants");
var useTypedSelector_1 = __importDefault(require("../../hooks/useTypedSelector"));
var TagView_1 = __importDefault(require("../TagView"));
var TagsPanel = function () {
    var _a;
    var tagsPanelVisible = useTypedSelector_1.default(function (state) { return state.tags.panelVisible; });
    if (!tagsPanelVisible) {
        return null;
    }
    return (react_1.default.createElement(ui_1.Box, { style: {
            position: 'relative',
            width: constants_1.TAGS_PANEL_WIDTH
        } },
        react_1.default.createElement(ui_1.Box, { className: "media__custom-scrollbar", style: {
                borderLeft: "1px solid " + ((_a = color_1.hues.gray) === null || _a === void 0 ? void 0 : _a[900].hex),
                height: '100%',
                overflowY: 'auto',
                position: 'absolute',
                right: 0,
                top: 0,
                width: '100%'
            } },
            react_1.default.createElement(TagView_1.default, null))));
};
exports.default = TagsPanel;
