// sanity-plugin-media/dist/components/DialogTagEdit/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var yup_1 = require("@hookform/resolvers/yup");
var ui_1 = require("@sanity/ui");
var groq_1 = __importDefault(require("groq"));
var react_1 = __importStar(require("react"));
var react_hook_form_1 = require("react-hook-form");
var react_redux_1 = require("react-redux");
var yup = __importStar(require("yup"));
var client_1 = require("../../client");
var constants_1 = require("../../constants");
var useTypedSelector_1 = __importDefault(require("../../hooks/useTypedSelector"));
var dialog_1 = require("../../modules/dialog");
var tags_1 = require("../../modules/tags");
var sanitizeFormData_1 = __importDefault(require("../../utils/sanitizeFormData"));
var Dialog_1 = __importDefault(require("../Dialog"));
var FormFieldInputText_1 = __importDefault(require("../FormFieldInputText"));
var FormSubmitButton_1 = __importDefault(require("../FormSubmitButton"));
var formSchema = yup.object().shape({
    name: yup.string().required('Name cannot be empty')
});
var DialogTagEdit = function (props) {
    var children = props.children, _a = props.dialog, id = _a.id, tagId = _a.tagId;
    // Redux
    var dispatch = react_redux_1.useDispatch();
    var tagItem = useTypedSelector_1.default(function (state) { return tags_1.selectTagById(state, String(tagId)); }); // TODO: double check string cast
    // State
    // - Generate a snapshot of the current tag
    var _b = __read(react_1.useState(tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag), 2), tagSnapshot = _b[0], setTagSnapshot = _b[1];
    var currentTag = tagItem ? tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag : tagSnapshot;
    var generateDefaultValues = function (tag) {
        var _a;
        return ({
            name: ((_a = tag === null || tag === void 0 ? void 0 : tag.name) === null || _a === void 0 ? void 0 : _a.current) || ''
        });
    };
    // react-hook-form
    var _c = react_hook_form_1.useForm({
        // defaultValues: {
        //   name: currentTag?.tag?.name?.current
        // },
        defaultValues: generateDefaultValues(tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag),
        mode: 'onChange',
        resolver: yup_1.yupResolver(formSchema)
    }), 
    // Read the formState before render to subscribe the form state through Proxy
    _d = _c.formState, errors = _d.errors, isDirty = _d.isDirty, isValid = _d.isValid, handleSubmit = _c.handleSubmit, register = _c.register, reset = _c.reset, setError = _c.setError;
    var formUpdating = !tagItem || (tagItem === null || tagItem === void 0 ? void 0 : tagItem.updating);
    // Callbacks
    var handleClose = function () {
        dispatch(dialog_1.dialogActions.remove({ id: id }));
    };
    // - submit react-hook-form
    var onSubmit = function (formData) { return __awaiter(void 0, void 0, void 0, function () {
        var sanitizedFormData;
        var _a;
        return __generator(this, function (_b) {
            if (!(tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag)) {
                return [2 /*return*/];
            }
            sanitizedFormData = sanitizeFormData_1.default(formData);
            dispatch(tags_1.tagsActions.updateRequest({
                closeDialogId: (_a = tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag) === null || _a === void 0 ? void 0 : _a._id,
                formData: {
                    name: {
                        _type: 'slug',
                        current: sanitizedFormData.name
                    }
                },
                tag: tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag
            }));
            return [2 /*return*/];
        });
    }); };
    var handleDelete = function () {
        var _a;
        if (!(tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag)) {
            return;
        }
        dispatch(dialog_1.dialogActions.showConfirmDeleteTag({
            closeDialogId: (_a = tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag) === null || _a === void 0 ? void 0 : _a._id,
            tag: tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag
        }));
    };
    var handleTagUpdate = function (update) {
        var result = update.result, transition = update.transition;
        if (result && transition === 'update') {
            // Regenerate snapshot
            setTagSnapshot(result);
            // Reset react-hook-form
            reset(generateDefaultValues(result));
        }
    };
    // Effects
    react_1.useEffect(function () {
        var _a;
        if (tagItem === null || tagItem === void 0 ? void 0 : tagItem.error) {
            setError('name', {
                message: (_a = tagItem.error) === null || _a === void 0 ? void 0 : _a.message
            });
        }
    }, [tagItem === null || tagItem === void 0 ? void 0 : tagItem.error]);
    // - Listen for asset mutations and update snapshot
    react_1.useEffect(function () {
        if (!(tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag)) {
            return undefined;
        }
        // Remember that Sanity listeners ignore joins, order clauses and projections
        var subscriptionAsset = client_1.client
            .listen(groq_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["*[_id == $id]"], ["*[_id == $id]"]))), { id: tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag._id })
            .subscribe(handleTagUpdate);
        return function () {
            subscriptionAsset === null || subscriptionAsset === void 0 ? void 0 : subscriptionAsset.unsubscribe();
        };
    }, []);
    var Footer = function () {
        var _a;
        return (react_1.default.createElement(ui_1.Box, { padding: 3 },
            react_1.default.createElement(ui_1.Flex, { justify: "space-between" },
                react_1.default.createElement(ui_1.Button, { disabled: formUpdating, fontSize: 1, mode: "bleed", onClick: handleDelete, text: "Delete", tone: "critical" }),
                react_1.default.createElement(FormSubmitButton_1.default, { disabled: formUpdating || !isDirty || !isValid, isValid: isValid, lastUpdated: (_a = tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag) === null || _a === void 0 ? void 0 : _a._updatedAt, onClick: handleSubmit(onSubmit) }))));
    };
    if (!currentTag) {
        return null;
    }
    return (react_1.default.createElement(Dialog_1.default, { footer: react_1.default.createElement(Footer, null), header: "Edit Tag", id: id, onClose: handleClose, width: 1, zOffset: constants_1.Z_INDEX_DIALOG },
        react_1.default.createElement(ui_1.Box, { as: "form", padding: 4, onSubmit: handleSubmit(onSubmit) },
            !tagItem && (react_1.default.createElement(ui_1.Card, { marginBottom: 3, padding: 3, radius: 2, shadow: 1, tone: "critical" },
                react_1.default.createElement(ui_1.Text, { size: 1 }, "This tag cannot be found \u2013 it may have been deleted."))),
            react_1.default.createElement("button", { style: { display: 'none' }, tabIndex: -1, type: "submit" }),
            react_1.default.createElement(FormFieldInputText_1.default, { disabled: formUpdating, error: errors === null || errors === void 0 ? void 0 : errors.name, label: "Name", name: "name", ref: register })),
        children));
};
exports.default = DialogTagEdit;
var templateObject_1;
