// sanity-plugin-media/dist/components/TextInputSearch/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var icons_1 = require("@sanity/icons");
var ui_1 = require("@sanity/ui");
var react_1 = __importDefault(require("react"));
var react_redux_1 = require("react-redux");
var useTypedSelector_1 = __importDefault(require("../../hooks/useTypedSelector"));
var search_1 = require("../../modules/search");
var TextInputSearch = function () {
    // Redux
    var searchQuery = useTypedSelector_1.default(function (state) { return state.search.query; });
    // Redux
    var dispatch = react_redux_1.useDispatch();
    // Callbacks
    var handleChange = function (e) {
        dispatch(search_1.searchActions.querySet({ searchQuery: e.currentTarget.value }));
    };
    var handleClear = function () {
        dispatch(search_1.searchActions.querySet({ searchQuery: '' }));
    };
    return (react_1.default.createElement(ui_1.Box, { style: { position: 'relative' } },
        react_1.default.createElement(ui_1.TextInput, { fontSize: 1, icon: icons_1.SearchIcon, onChange: handleChange, placeholder: "Search", radius: 2, value: searchQuery }),
        searchQuery.length > 0 && (react_1.default.createElement(ui_1.Flex, { align: "center", justify: "center", onClick: handleClear, style: {
                cursor: 'pointer',
                height: '100%',
                opacity: 0.75,
                position: 'absolute',
                right: 0,
                top: 0,
                width: '2em',
                zIndex: 1 // force stacking context
            } },
            react_1.default.createElement(icons_1.CloseIcon, null)))));
};
exports.default = TextInputSearch;
