// sanity-plugin-media/dist/components/CardUpload/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var color_1 = require("@sanity/color");
var icons_1 = require("@sanity/icons");
var ui_1 = require("@sanity/ui");
var filesize_1 = __importDefault(require("filesize"));
var react_1 = __importDefault(require("react"));
var react_redux_1 = require("react-redux");
var constants_1 = require("../../constants");
var useTypedSelector_1 = __importDefault(require("../../hooks/useTypedSelector"));
var uploads_1 = require("../../modules/uploads");
var FileIcon_1 = __importDefault(require("../FileIcon"));
var Image_1 = __importDefault(require("../Image"));
var CardUpload = function (props) {
    var id = props.id;
    // Redux
    var dispatch = react_redux_1.useDispatch();
    var item = useTypedSelector_1.default(function (state) { return uploads_1.selectUploadById(state, id); });
    if (!item) {
        return null;
    }
    var fileSize = filesize_1.default(item.size, { base: 10, round: 0 });
    var percentLoaded = Math.round(item.percent || 0); // (0 - 100)
    var isComplete = item.status === 'complete';
    var isUploading = item.status === 'uploading';
    var isQueued = item.status === 'queued';
    var status;
    if (isComplete) {
        status = 'Verifying';
    }
    if (isUploading) {
        status = percentLoaded + "%";
    }
    if (isQueued) {
        status = 'Queued';
    }
    // Callbacks
    var handleCancelUpload = function () {
        dispatch(uploads_1.uploadsActions.uploadCancel({ hash: item.hash }));
    };
    return (react_1.default.createElement(ui_1.Flex, { direction: "column", style: {
            background: color_1.hues.gray[950].hex,
            border: '1px solid transparent',
            height: '100%',
            position: 'relative'
        } },
        react_1.default.createElement("div", { style: {
                background: color_1.hues.gray[600].hex,
                bottom: 0,
                height: '1px',
                left: 0,
                position: 'absolute',
                width: '100%',
                transform: "scaleX(" + percentLoaded * 0.01 + ")",
                transformOrigin: 'bottom left',
                transition: 'all 1000ms ease-out'
            } }),
        react_1.default.createElement(ui_1.Box, { flex: 1, style: { position: 'relative' } },
            item.assetType === 'image' && (item === null || item === void 0 ? void 0 : item.objectUrl) && (react_1.default.createElement(Image_1.default, { draggable: false, src: item.objectUrl, style: {
                    opacity: 0.4
                } })),
            item.assetType === 'file' && (react_1.default.createElement("div", { style: { height: '100%', opacity: 0.1 } },
                react_1.default.createElement(FileIcon_1.default, { width: "80px" }))),
            !isComplete && percentLoaded !== 100 && (react_1.default.createElement(ui_1.Flex, { align: "center", direction: "column", justify: "center", style: {
                    height: '100%',
                    left: 0,
                    position: 'absolute',
                    top: 0,
                    width: '100%'
                } },
                react_1.default.createElement(ui_1.Tooltip, { content: react_1.default.createElement(ui_1.Box, { padding: 2 },
                        react_1.default.createElement(ui_1.Text, { muted: true, size: 1 }, "Cancel")), disabled: 'ontouchstart' in window, placement: "top" },
                    react_1.default.createElement(ui_1.Button, { fontSize: 4, icon: icons_1.CloseIcon, mode: "bleed", onClick: handleCancelUpload, padding: 2, style: { background: 'none', boxShadow: 'none' }, tone: "critical" }))))),
        react_1.default.createElement(ui_1.Flex, { align: "center", justify: "space-between", paddingX: 2, style: { height: constants_1.PANEL_HEIGHT + "px" } },
            react_1.default.createElement(ui_1.Box, { flex: 1, marginRight: 1 },
                react_1.default.createElement(ui_1.Text, { size: 0, textOverflow: "ellipsis" },
                    item.name,
                    " (",
                    fileSize,
                    ")")),
            react_1.default.createElement(ui_1.Text, { size: 0, style: { flexShrink: 0 }, weight: "semibold" }, status))));
};
exports.default = CardUpload;
