// sanity-plugin-media/dist/app.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ui_1 = require("@sanity/ui");
var react_1 = __importStar(require("react"));
var theme_ui_1 = require("theme-ui");
var Browser_1 = __importDefault(require("./components/Browser"));
var ReduxProvider_1 = __importDefault(require("./components/ReduxProvider"));
var constants_1 = require("./constants");
var AssetSourceDispatchContext_1 = require("./contexts/AssetSourceDispatchContext");
var useKeyPress_1 = __importDefault(require("./hooks/useKeyPress"));
var GlobalStyles_1 = __importDefault(require("./styled/GlobalStyles"));
var theme_1 = __importDefault(require("./styled/theme"));
var AssetBrowser = react_1.forwardRef(function (props, ref) {
    var onClose = props.onClose, onSelect = props.onSelect;
    // Close on escape key press
    useKeyPress_1.default('escape', onClose);
    // Stop propagation and prevent document mouse events from firing.
    // This is a bit of a hack to make this work with `editModal = 'popover'` and prevent Sanity's <Popover /> component from
    // prematurely closing, as it attaches events on `document` to detect outside clicks.
    var handleStopPropagation = function (e) {
        e.nativeEvent.stopImmediatePropagation();
        e.stopPropagation();
    };
    return (react_1.default.createElement(ReduxProvider_1.default, __assign({}, props),
        react_1.default.createElement(ui_1.ThemeProvider, { scheme: "dark", theme: ui_1.studioTheme },
            react_1.default.createElement(theme_ui_1.ThemeProvider, { theme: theme_1.default },
                react_1.default.createElement(ui_1.PortalProvider, { element: document.body },
                    react_1.default.createElement(ui_1.ToastProvider, { zOffset: constants_1.Z_INDEX_TOAST_PROVIDER },
                        react_1.default.createElement(AssetSourceDispatchContext_1.AssetBrowserDispatchProvider, { onSelect: onSelect },
                            react_1.default.createElement(GlobalStyles_1.default, null),
                            onClose ? (react_1.default.createElement(ui_1.Portal, null,
                                react_1.default.createElement(ui_1.Box, { onDragEnter: handleStopPropagation, onDragLeave: handleStopPropagation, onDragOver: handleStopPropagation, onDrop: handleStopPropagation, onMouseUp: handleStopPropagation, ref: ref, style: {
                                        bottom: 0,
                                        height: 'auto',
                                        left: 0,
                                        position: 'fixed',
                                        top: 0,
                                        width: '100%',
                                        zIndex: constants_1.Z_INDEX_APP
                                    } },
                                    react_1.default.createElement(Browser_1.default, { onClose: onClose })))) : (react_1.default.createElement(ui_1.Flex, { direction: "column", ref: ref },
                                react_1.default.createElement(Browser_1.default, null))))))))));
});
exports.default = AssetBrowser;
