// sanity-plugin-media/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Z_INDEX_TOAST_PROVIDER = exports.Z_INDEX_DIALOG = exports.Z_INDEX_APP = exports.TAGS_PANEL_WIDTH = exports.TAG_DOCUMENT_NAME = exports.PANEL_HEIGHT = exports.API_VERSION = exports.FACETS = exports.ORDER_OPTIONS = void 0;
var searchFacets_1 = require("./config/searchFacets");
// Sort order dropdown options
// null values are represented as menu dividers
exports.ORDER_OPTIONS = [
    {
        direction: 'desc',
        field: '_createdAt'
    },
    {
        direction: 'asc',
        field: '_createdAt'
    },
    // Divider
    null,
    {
        direction: 'desc',
        field: '_updatedAt'
    },
    {
        direction: 'asc',
        field: '_updatedAt'
    },
    // Divider
    null,
    {
        direction: 'asc',
        field: 'originalFilename'
    },
    {
        direction: 'desc',
        field: 'originalFilename'
    },
    // Divider
    null,
    {
        direction: 'desc',
        field: 'size'
    },
    {
        direction: 'asc',
        field: 'size'
    }
];
exports.FACETS = [
    searchFacets_1.inputs.tag,
    searchFacets_1.divider,
    searchFacets_1.inputs.inUse,
    searchFacets_1.inputs.inCurrentDocument,
    searchFacets_1.divider,
    searchFacets_1.inputs.title,
    searchFacets_1.inputs.altText,
    searchFacets_1.inputs.description,
    searchFacets_1.divider,
    searchFacets_1.inputs.isOpaque,
    searchFacets_1.divider,
    searchFacets_1.inputs.fileName,
    searchFacets_1.inputs.size,
    searchFacets_1.inputs.type,
    searchFacets_1.divider,
    searchFacets_1.inputs.orientation,
    searchFacets_1.inputs.width,
    searchFacets_1.inputs.height
];
exports.API_VERSION = '2021-06-07';
exports.PANEL_HEIGHT = 32; // px
exports.TAG_DOCUMENT_NAME = 'media.tag';
exports.TAGS_PANEL_WIDTH = 250; // px
// NOTE: Manually set plugin z-index values to be higher than Sanity's header search field
// (which is currently 500202). Also ensure toasts always sit above dialogs.
exports.Z_INDEX_APP = 600000;
exports.Z_INDEX_DIALOG = 600001;
exports.Z_INDEX_TOAST_PROVIDER = 600002;
