// sanity-plugin-media/dist/components/DialogAssetEdit/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var yup_1 = require("@hookform/resolvers/yup");
var ui_1 = require("@sanity/ui");
var groq_1 = __importDefault(require("groq"));
var react_1 = __importStar(require("react"));
var react_hook_form_1 = require("react-hook-form");
var react_redux_1 = require("react-redux");
var theme_ui_1 = require("theme-ui");
var yup = __importStar(require("yup"));
var client_1 = require("../../client");
var constants_1 = require("../../constants");
var useTypedSelector_1 = __importDefault(require("../../hooks/useTypedSelector"));
var assets_1 = require("../../modules/assets");
var dialog_1 = require("../../modules/dialog");
var tags_1 = require("../../modules/tags");
var getTagSelectOptions_1 = __importDefault(require("../../utils/getTagSelectOptions"));
var imageDprUrl_1 = __importDefault(require("../../utils/imageDprUrl"));
var sanitizeFormData_1 = __importDefault(require("../../utils/sanitizeFormData"));
var typeGuards_1 = require("../../utils/typeGuards");
var AssetMetadata_1 = __importDefault(require("../AssetMetadata"));
var Dialog_1 = __importDefault(require("../Dialog"));
var DocumentList_1 = __importDefault(require("../DocumentList"));
var FileAssetPreview_1 = __importDefault(require("../FileAssetPreview"));
var FormFieldInputFilename_1 = __importDefault(require("../FormFieldInputFilename"));
var FormFieldInputTags_1 = __importDefault(require("../FormFieldInputTags"));
var FormFieldInputText_1 = __importDefault(require("../FormFieldInputText"));
var FormFieldInputTextarea_1 = __importDefault(require("../FormFieldInputTextarea"));
var FormSubmitButton_1 = __importDefault(require("../FormSubmitButton"));
var Image_1 = __importDefault(require("../Image"));
var formSchema = yup.object().shape({
    originalFilename: yup.string().required('Filename cannot be empty')
});
var getFilenameWithoutExtension = function (asset) {
    var _a, _b;
    var extensionIndex = (_a = asset === null || asset === void 0 ? void 0 : asset.originalFilename) === null || _a === void 0 ? void 0 : _a.lastIndexOf("." + asset.extension);
    return (_b = asset === null || asset === void 0 ? void 0 : asset.originalFilename) === null || _b === void 0 ? void 0 : _b.slice(0, extensionIndex);
};
var DialogAssetEdit = function (props) {
    var _a, _b, _c;
    var children = props.children, _d = props.dialog, assetId = _d.assetId, id = _d.id, lastCreatedTag = _d.lastCreatedTag, lastRemovedTagIds = _d.lastRemovedTagIds;
    // Redux
    var dispatch = react_redux_1.useDispatch();
    var assetItem = useTypedSelector_1.default(function (state) { return assets_1.selectAssetById(state, String(assetId)); }); // TODO: check casting
    var tags = useTypedSelector_1.default(tags_1.selectTags);
    // Refs
    var isMounted = react_1.useRef(false);
    // State
    // - Generate a snapshot of the current asset
    var _e = __read(react_1.useState(assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset), 2), assetSnapshot = _e[0], setAssetSnapshot = _e[1];
    var _f = __read(react_1.useState('details'), 2), tabSection = _f[0], setTabSection = _f[1];
    var currentAsset = assetItem ? assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset : assetSnapshot;
    var allTagOptions = getTagSelectOptions_1.default(tags);
    // Redux
    var assetTagOptions = useTypedSelector_1.default(tags_1.selectTagSelectOptions(currentAsset));
    var generateDefaultValues = function (asset) { return ({
        altText: (asset === null || asset === void 0 ? void 0 : asset.altText) || '',
        description: (asset === null || asset === void 0 ? void 0 : asset.description) || '',
        originalFilename: asset ? getFilenameWithoutExtension(asset) : undefined,
        opt: { media: { tags: assetTagOptions } },
        title: (asset === null || asset === void 0 ? void 0 : asset.title) || ''
    }); };
    // Generate a string from all current tag labels
    // This is used purely to determine tag updates to then update the form in real time
    var currentTagLabels = assetTagOptions === null || assetTagOptions === void 0 ? void 0 : assetTagOptions.map(function (tag) { return tag.label; }).join(',');
    // react-hook-form
    var _g = react_hook_form_1.useForm({
        defaultValues: generateDefaultValues(assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset),
        mode: 'onChange',
        resolver: yup_1.yupResolver(formSchema)
    }), control = _g.control, 
    // Read the formState before render to subscribe the form state through Proxy
    _h = _g.formState, errors = _h.errors, isDirty = _h.isDirty, isValid = _h.isValid, getValues = _g.getValues, handleSubmit = _g.handleSubmit, register = _g.register, reset = _g.reset, setValue = _g.setValue;
    var formUpdating = !assetItem || (assetItem === null || assetItem === void 0 ? void 0 : assetItem.updating);
    // Callbacks
    var handleClose = function () {
        dispatch(dialog_1.dialogActions.remove({ id: id }));
    };
    var handleDelete = function () {
        if (!(assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset)) {
            return;
        }
        dispatch(dialog_1.dialogActions.showConfirmDeleteAssets({
            assets: [assetItem],
            closeDialogId: assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset._id
        }));
    };
    var handleAssetUpdate = function (update) {
        var result = update.result, transition = update.transition;
        if (result && transition === 'update') {
            // Regenerate asset snapshot
            setAssetSnapshot(result);
            // Reset react-hook-form
            reset(generateDefaultValues(result));
        }
    };
    var handleCreateTag = function (tagName) {
        // Dispatch action to create new tag
        dispatch(tags_1.tagsActions.createRequest({
            assetId: currentAsset === null || currentAsset === void 0 ? void 0 : currentAsset._id,
            name: tagName
        }));
    };
    // - submit react-hook-form
    var onSubmit = function (formData) { return __awaiter(void 0, void 0, void 0, function () {
        var sanitizedFormData;
        var _a;
        return __generator(this, function (_b) {
            if (!(assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset)) {
                return [2 /*return*/];
            }
            sanitizedFormData = sanitizeFormData_1.default(formData);
            dispatch(assets_1.assetsActions.updateRequest({
                asset: assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset,
                closeDialogId: assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset._id,
                formData: __assign(__assign({}, sanitizedFormData), { 
                    // Map tags to sanity references
                    opt: {
                        media: __assign(__assign({}, sanitizedFormData.opt.media), { tags: ((_a = sanitizedFormData.opt.media.tags) === null || _a === void 0 ? void 0 : _a.map(function (tag) { return ({
                                _ref: tag.value,
                                _type: 'reference',
                                _weak: true
                            }); })) || null })
                    }, 
                    // Append extension to filename
                    originalFilename: sanitizedFormData.originalFilename + "." + (assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset.extension) })
            }));
            return [2 /*return*/];
        });
    }); };
    // Effects
    // - Listen for asset mutations and update snapshot
    react_1.useEffect(function () {
        if (!(assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset)) {
            return undefined;
        }
        // Remember that Sanity listeners ignore joins, order clauses and projections
        var subscriptionAsset = client_1.client
            .listen(groq_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["*[_id == $id]"], ["*[_id == $id]"]))), { id: assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset._id })
            .subscribe(handleAssetUpdate);
        return function () {
            subscriptionAsset === null || subscriptionAsset === void 0 ? void 0 : subscriptionAsset.unsubscribe();
        };
    }, []);
    // - Partially reset form when current tags have changed (and after initial mount)
    react_1.useEffect(function () {
        if (isMounted.current) {
            reset({
                opt: {
                    media: { tags: assetTagOptions }
                }
            }, {
                errors: true,
                dirtyFields: true,
                isDirty: true
            });
        }
        // Mark as mounted
        isMounted.current = true;
    }, [currentTagLabels]);
    // - Update tags form field (react-select) when a new _inline_ tag has been created
    react_1.useEffect(function () {
        if (lastCreatedTag) {
            var existingTags = getValues('opt.media.tags') || [];
            var updatedTags = existingTags.concat([lastCreatedTag]);
            setValue('opt.media.tags', updatedTags, { shouldDirty: true });
        }
    }, [lastCreatedTag]);
    // - Update tags form field (react-select) when an _inline_ tag has been removed elsewhere
    react_1.useEffect(function () {
        if (lastRemovedTagIds) {
            var existingTags = getValues('opt.media.tags') || [];
            var updatedTags = existingTags.filter(function (tag) {
                return !lastRemovedTagIds.includes(tag.value);
            });
            setValue('opt.media.tags', updatedTags, { shouldDirty: true });
        }
    }, [lastRemovedTagIds]);
    var Footer = function () { return (react_1.default.createElement(ui_1.Box, { padding: 3 },
        react_1.default.createElement(ui_1.Flex, { justify: "space-between" },
            react_1.default.createElement(ui_1.Button, { disabled: formUpdating, fontSize: 1, mode: "bleed", onClick: handleDelete, text: "Delete", tone: "critical" }),
            react_1.default.createElement(FormSubmitButton_1.default, { disabled: formUpdating || !isDirty || !isValid, isValid: isValid, lastUpdated: currentAsset._updatedAt, onClick: handleSubmit(onSubmit) })))); };
    if (!currentAsset) {
        return null;
    }
    return (react_1.default.createElement(Dialog_1.default, { footer: react_1.default.createElement(Footer, null), header: "Asset details", id: id, onClose: handleClose, width: 3, zOffset: constants_1.Z_INDEX_DIALOG },
        react_1.default.createElement(ui_1.Flex, { direction: ['column-reverse', 'column-reverse', 'row-reverse'] },
            react_1.default.createElement(ui_1.Box, { flex: 1, marginTop: [5, 5, 0], padding: 4 },
                react_1.default.createElement(ui_1.TabList, { space: 2 },
                    react_1.default.createElement(ui_1.Tab, { "aria-controls": "details-panel", disabled: formUpdating, id: "details-tab", label: "Details", onClick: function () { return setTabSection('details'); }, selected: tabSection === 'details', size: 2 }),
                    react_1.default.createElement(ui_1.Tab, { "aria-controls": "references-panel", disabled: formUpdating, id: "references-tab", label: "References", onClick: function () { return setTabSection('references'); }, selected: tabSection === 'references', size: 2 })),
                react_1.default.createElement(ui_1.Box, { as: "form", marginTop: 4, onSubmit: handleSubmit(onSubmit) },
                    !assetItem && (react_1.default.createElement(ui_1.Card, { marginBottom: 3, padding: 3, radius: 2, shadow: 1, tone: "critical" },
                        react_1.default.createElement(ui_1.Text, { size: 1 }, "This file cannot be found \u2013 it may have been deleted."))),
                    react_1.default.createElement("button", { style: { display: 'none' }, tabIndex: -1, type: "submit" }),
                    react_1.default.createElement(ui_1.TabPanel, { "aria-labelledby": "details", hidden: tabSection !== 'details', id: "details-panel" },
                        react_1.default.createElement(ui_1.Stack, { space: 3 },
                            react_1.default.createElement(FormFieldInputTags_1.default, { control: control, disabled: formUpdating, error: (_b = (_a = errors === null || errors === void 0 ? void 0 : errors.opt) === null || _a === void 0 ? void 0 : _a.media) === null || _b === void 0 ? void 0 : _b.tags, label: "Tags", name: "opt.media.tags", onCreateTag: handleCreateTag, options: allTagOptions, placeholder: "Select or create...", value: assetTagOptions }),
                            react_1.default.createElement(FormFieldInputFilename_1.default, { disabled: formUpdating, error: errors === null || errors === void 0 ? void 0 : errors.originalFilename, extension: (currentAsset === null || currentAsset === void 0 ? void 0 : currentAsset.extension) || '', label: "Filename", name: "originalFilename", ref: register, value: getFilenameWithoutExtension(currentAsset) }),
                            react_1.default.createElement(FormFieldInputText_1.default, { disabled: formUpdating, error: errors === null || errors === void 0 ? void 0 : errors.title, label: "Title", name: "title", ref: register, value: currentAsset === null || currentAsset === void 0 ? void 0 : currentAsset.title }),
                            react_1.default.createElement(FormFieldInputText_1.default, { disabled: formUpdating, error: errors === null || errors === void 0 ? void 0 : errors.altText, label: "Alt Text", name: "altText", ref: register, value: currentAsset === null || currentAsset === void 0 ? void 0 : currentAsset.altText }),
                            react_1.default.createElement(FormFieldInputTextarea_1.default, { disabled: formUpdating, error: errors === null || errors === void 0 ? void 0 : errors.description, label: "Description", name: "description", ref: register, rows: 3, value: currentAsset === null || currentAsset === void 0 ? void 0 : currentAsset.description }))),
                    react_1.default.createElement(ui_1.TabPanel, { "aria-labelledby": "references", hidden: tabSection !== 'references', id: "references-panel" },
                        react_1.default.createElement(ui_1.Box, { marginTop: 5 }, (assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset) && react_1.default.createElement(DocumentList_1.default, { assetId: assetItem === null || assetItem === void 0 ? void 0 : assetItem.asset._id }))))),
            react_1.default.createElement(ui_1.Box, { flex: 1, padding: 4 },
                react_1.default.createElement(theme_ui_1.AspectRatio, { ratio: 1 },
                    typeGuards_1.isFileAsset(currentAsset) && react_1.default.createElement(FileAssetPreview_1.default, { asset: currentAsset }),
                    typeGuards_1.isImageAsset(currentAsset) && (react_1.default.createElement(Image_1.default, { draggable: false, showCheckerboard: !((_c = currentAsset === null || currentAsset === void 0 ? void 0 : currentAsset.metadata) === null || _c === void 0 ? void 0 : _c.isOpaque), src: imageDprUrl_1.default(currentAsset, { height: 600, width: 600 }) }))),
                currentAsset && (react_1.default.createElement(ui_1.Box, { marginTop: 4 },
                    react_1.default.createElement(AssetMetadata_1.default, { asset: currentAsset, item: assetItem }))))),
        children));
};
exports.default = DialogAssetEdit;
var templateObject_1;
