// sanity-plugin-media/dist/components/Progress/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ui_1 = require("@sanity/ui");
var react_nprogress_1 = require("@tanem/react-nprogress");
var react_1 = __importDefault(require("react"));
var Progress = function (props) {
    var loading = props.loading;
    var _a = react_nprogress_1.useNProgress({
        animationDuration: 300,
        isAnimating: loading
    }), animationDuration = _a.animationDuration, isFinished = _a.isFinished, progress = _a.progress;
    return (react_1.default.createElement(ui_1.Box, { style: {
            opacity: isFinished ? 0 : 1,
            transition: "opacity " + animationDuration + "ms linear"
        } },
        react_1.default.createElement(ui_1.Box, { style: {
                height: '1px',
                background: 'rgba(255, 255, 255, 0.5)',
                position: 'absolute',
                left: 0,
                top: 0,
                transition: "width " + animationDuration + "ms linear",
                width: progress * 100 + "%"
            } })));
};
exports.default = Progress;
