// sanity-plugin-media/dist/components/TableHeader/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
var color_1 = require("@sanity/color");
var ui_1 = require("@sanity/ui");
var react_1 = __importDefault(require("react"));
var react_redux_1 = require("react-redux");
var styled_components_1 = __importDefault(require("styled-components"));
var theme_ui_1 = require("theme-ui");
var constants_1 = require("../../constants");
var AssetSourceDispatchContext_1 = require("../../contexts/AssetSourceDispatchContext");
var useTypedSelector_1 = __importDefault(require("../../hooks/useTypedSelector"));
var assets_1 = require("../../modules/assets");
var TableHeaderItem_1 = __importDefault(require("../TableHeaderItem"));
// TODO: DRY
var ContextActionContainer = styled_components_1.default(ui_1.Flex)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  cursor: pointer;\n\n  @media (hover: hover) and (pointer: fine) {\n    &:hover {\n      background: ", ";\n    }\n  }\n"], ["\n  cursor: pointer;\n\n  @media (hover: hover) and (pointer: fine) {\n    &:hover {\n      background: ", ";\n    }\n  }\n"])), (_a = color_1.hues.gray) === null || _a === void 0 ? void 0 : _a[900].hex);
var TableHeader = function () {
    var _a;
    // Redux
    var dispatch = react_redux_1.useDispatch();
    var fetching = useTypedSelector_1.default(function (state) { return state.assets.fetching; });
    var itemsLength = useTypedSelector_1.default(assets_1.selectAssetsLength);
    var numPickedAssets = useTypedSelector_1.default(assets_1.selectAssetsPickedLength);
    var onSelect = AssetSourceDispatchContext_1.useAssetSourceActions().onSelect;
    var allSelected = numPickedAssets === itemsLength;
    // Callbacks
    var handleContextActionClick = function (e) {
        e.stopPropagation();
        if (allSelected) {
            dispatch(assets_1.assetsActions.pickClear());
        }
        else {
            dispatch(assets_1.assetsActions.pickAll());
        }
    };
    // Note that even though we hide the table header on smaller breakpoints, we never set it to
    // `display: none`, as doing so causes issues with react-virtuoso.
    // Instead, we give it 0 height and hide it with `visibility: hidden`.
    return (react_1.default.createElement(theme_ui_1.Box, { sx: {
            alignItems: 'center',
            bg: color_1.black.hex,
            borderBottom: "1px solid " + ((_a = color_1.hues.gray) === null || _a === void 0 ? void 0 : _a[900].hex),
            display: 'grid',
            gridColumnGap: [0, null, null, 3],
            gridTemplateColumns: 'tableLarge',
            height: [0, null, null, constants_1.PANEL_HEIGHT + "px"],
            letterSpacing: '0.025em',
            position: 'sticky',
            textTransform: 'uppercase',
            top: 0,
            visibility: ['hidden', null, null, 'visible'],
            width: '100%',
            zIndex: 1 // force stacking context
        } },
        onSelect ? (react_1.default.createElement(TableHeaderItem_1.default, null)) : (react_1.default.createElement(ContextActionContainer, { align: "center", justify: "center", onClick: handleContextActionClick, style: {
                height: '100%',
                position: 'relative'
            } },
            react_1.default.createElement(ui_1.Checkbox, { checked: !fetching && allSelected, readOnly: true, style: {
                    pointerEvents: 'none',
                    transform: 'scale(0.8)'
                } }))),
        react_1.default.createElement(TableHeaderItem_1.default, null),
        react_1.default.createElement(TableHeaderItem_1.default, { field: "originalFilename", title: "Filename" }),
        react_1.default.createElement(TableHeaderItem_1.default, { title: "Resolution" }),
        react_1.default.createElement(TableHeaderItem_1.default, { field: "mimeType", title: "MIME type" }),
        react_1.default.createElement(TableHeaderItem_1.default, { field: "size", title: "Size" }),
        react_1.default.createElement(TableHeaderItem_1.default, { field: "_updatedAt", title: "Last updated" }),
        react_1.default.createElement(TableHeaderItem_1.default, null)));
};
exports.default = TableHeader;
var templateObject_1;
