// sanity-plugin-media/dist/modules/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rootReducer = exports.rootEpic = void 0;
var toolkit_1 = require("@reduxjs/toolkit");
var redux_observable_1 = require("redux-observable");
var assets_1 = __importStar(require("./assets"));
var debug_1 = __importDefault(require("./debug"));
var dialog_1 = __importStar(require("./dialog"));
var selected_1 = __importDefault(require("./selected"));
var notifications_1 = __importStar(require("./notifications"));
var search_1 = __importStar(require("./search"));
var tags_1 = __importStar(require("./tags"));
var uploads_1 = __importStar(require("./uploads"));
exports.rootEpic = redux_observable_1.combineEpics(assets_1.assetsDeleteEpic, assets_1.assetsFetchEpic, assets_1.assetsFetchAfterDeleteAllEpic, assets_1.assetsFetchNextPageEpic, assets_1.assetsFetchPageIndexEpic, assets_1.assetsListenerCreateQueueEpic, assets_1.assetsListenerDeleteQueueEpic, assets_1.assetsListenerUpdateQueueEpic, assets_1.assetsOrderSetEpic, assets_1.assetsSearchEpic, assets_1.assetsSortEpic, assets_1.assetsTagsAddEpic, assets_1.assetsTagsRemoveEpic, assets_1.assetsUnpickEpic, assets_1.assetsUpdateEpic, dialog_1.dialogClearOnAssetUpdateEpic, dialog_1.dialogTagCreateEpic, dialog_1.dialogTagDeleteEpic, notifications_1.notificationsAssetsDeleteErrorEpic, notifications_1.notificationsAssetsDeleteCompleteEpic, notifications_1.notificationsAssetsTagsAddCompleteEpic, notifications_1.notificationsAssetsTagsRemoveCompleteEpic, notifications_1.notificationsAssetsUpdateCompleteEpic, notifications_1.notificationsGenericErrorEpic, notifications_1.notificationsTagCreateCompleteEpic, notifications_1.notificationsTagDeleteCompleteEpic, notifications_1.notificationsTagUpdateCompleteEpic, search_1.searchFacetTagUpdateEpic, tags_1.tagsCreateEpic, tags_1.tagsDeleteEpic, tags_1.tagsFetchEpic, tags_1.tagsListenerCreateQueueEpic, tags_1.tagsListenerDeleteQueueEpic, tags_1.tagsListenerUpdateQueueEpic, tags_1.tagsSortEpic, tags_1.tagsUpdateEpic, uploads_1.uploadsAssetStartEpic, uploads_1.uploadsAssetUploadEpic, uploads_1.uploadsCheckRequestEpic, uploads_1.uploadsCompleteQueueEpic);
exports.rootReducer = toolkit_1.combineReducers({
    assets: assets_1.default,
    debug: debug_1.default,
    dialog: dialog_1.default,
    notifications: notifications_1.default,
    search: search_1.default,
    selected: selected_1.default,
    tags: tags_1.default,
    uploads: uploads_1.default
});
